File: quattro2/mac/Common/JavaScriptInterface/NativeCall+recorder.m
//
//  NativeCall+recorder.m
//
//  Copyright 2015 Roland Corporation. All rights reserved.
//

#import "AVFAudio/AVAudioApplication.h"
#import "AVFAudio/AVAudioSession.h"
#import "NativeCall.h"

static NSString *const OBJ_NAME = @"recorder";

@interface NativeCall (recorder) @end

@implementation NativeCall (recorder)

- (NSString *)$$recorder_device:(NSString *)param1 :(NSString *)param2
{
	if (param1) {
		recorder.device = param1;
	}
	return_s(recorder.device);
}

- (NSString *)$$recorder_create:(NSString *)param1 :(NSString *)param2
{
	return_b([recorder create:[NSURL fileURLWithPath:param1]
                       format:(enum RecordingFormat)[param2 intValue]]);
}

- (NSString *)$$recorder_record: (NSString *)param1 :(NSString *)param2
{
	void (^handler)(BOOL granted) = ^(BOOL granted) {
		if (granted) {
			[recorder record];
		} else {
			NSString *event = [NSString stringWithFormat:@"%@\f%@", OBJ_NAME, @"unauthorized"];
			[self postEvent:event];
		}
	};

#if TARGET_OS_IPHONE
	if (@available(iOS 17.0, *)) {
#elif TARGET_OS_MAC
	if (@available(macOS 14.0, *)) {
#endif
		[AVAudioApplication requestRecordPermissionWithCompletionHandler:handler];
	} else {
#if TARGET_OS_IPHONE
		[[AVAudioSession sharedInstance] requestRecordPermission:handler];
#elif TARGET_OS_MAC
		[recorder record];
#endif
	}
	undefined;
}

- (NSString *)$$recorder_pause:(NSString *)param1 :(NSString *)param2
{ [recorder pause]; undefined; }

- (NSString *)$$recorder_stop: (NSString *)param1 :(NSString *)param2
{ [recorder stop:[param1 boolValue]]; undefined; }

- (NSString *)$$recorder_file:(NSString *)param1 :(NSString *)param2
{ return_s([recorder.file path]); }

- (NSString *)$$recorder_status:(NSString *)param1 :(NSString *)param2
{ return_d(recorder.status); }

- (NSString *)$$recorder_channels:(NSString *)param1 :(NSString *)param2
{ return_u(recorder.numberOfChannels); }

- (NSString *)$$recorder_peakpower:(NSString *)param1 :(NSString *)param2
{ return_f([recorder peakPowerForChannel:[param1 intValue]]); }

- (NSString *)$$recorder_time:(NSString *)param1 :(NSString *)param2
{ return_f(recorder.currentTime); }

- (NSString *)$$recorder_volume:(NSString *)param1 :(NSString *)param2
{
	if (param1) {
		recorder.volume = [param1 floatValue];
	}
	return_f(recorder.volume);
}

- (NSString *)$$recorder_input:(NSString *)param1 :(NSString *)param2
{ recorder.input = [param1 boolValue] ? rwc.inputStream : input; undefined; }

@end

@implementation _AudioRecorderDelegate

- (void)audioRecorderDidFinishRecording:(NSURL *)fileURL
{
	NSString *event = [NSString stringWithFormat:@"%@\f%@\f%@", OBJ_NAME, @"stop", [fileURL path]];
    [self.native postEvent:event];
}

- (void)audioRecorderErrorDidOccur:(NSURL *)fileURL
{
	NSString *event = [NSString stringWithFormat:@"%@\f%@\f%@", OBJ_NAME, @"error", [fileURL path]];
    [self.native postEvent:event];
}

@end

File: quattro2/win10/JavaScriptInterface/NativeCall+recorder.cpp
//
//  NativeCall+recorder.cpp
//
//  Copyright 2015 Roland Corporation. All rights reserved.
//

#include "NativeCall.h"

static LPCWSTR OBJ_NAME = L"recorder";

enum {
	dispid_recorder_device,
	dispid_recorder_create,
	dispid_recorder_record,
	dispid_recorder_pause,
	dispid_recorder_stop,
	dispid_recorder_volume,
	dispid_recorder_file,
	dispid_recorder_status,
	dispid_recorder_channels,
	dispid_recorder_peakpower,
	dispid_recorder_time,
	dispid_recorder_input,
};

NativeCall_recorder::NativeCall_recorder(NativeObjects* objs) : NativeCall(objs)
{
	set(L"device",    dispid_recorder_device);
	set(L"create",    dispid_recorder_create);
	set(L"record",    dispid_recorder_record);
	set(L"pause",     dispid_recorder_pause);
	set(L"stop",      dispid_recorder_stop);
	set(L"volume",    dispid_recorder_volume);
	set(L"file",      dispid_recorder_file);
	set(L"status",    dispid_recorder_status);
	set(L"channels",  dispid_recorder_channels);
	set(L"peakpower", dispid_recorder_peakpower);
	set(L"time",      dispid_recorder_time);
	set(L"input",     dispid_recorder_input);
}

LPCWSTR NativeCall_recorder::InterfaceName() const { return OBJ_NAME; }

void NativeCall_recorder::Dispatch(UINT dispid, const CefV8ValueList& args, CefRetValue& ret)
{
	AudioRecorder* recorder = mp_objs->recorder;

	switch (dispid) {

		case dispid_recorder_device:
		{
			if (args.size() > 0) {
				recorder->device(CString(args[0]->GetStringValue().c_str()));
			}
			ret = recorder->device();
			break;
		}
		case dispid_recorder_create:
		{
			ret = recorder->create(CString(args[0]->GetStringValue().c_str()), (RecordingFormat)args[1]->GetIntValue());
			break;
		}
		case dispid_recorder_record:
		{
			recorder->record();
			break;
		}
		case dispid_recorder_pause:
		{
			recorder->pause();
			break;
		}
		case dispid_recorder_stop:
		{
			recorder->stop(false);
			break;
		}
		case dispid_recorder_volume:
		{
			if (args.size() > 0) {
				recorder->volume((float)args[0]->GetDoubleValue());
			}
			ret = recorder->volume();
			break;
		}
		case dispid_recorder_file:
		{
			ret = (recorder->file() ? CStringW(recorder->file()) : L"");
			break;
		}
		case dispid_recorder_status:
		{
			ret = recorder->status();
			break;
		}
		case dispid_recorder_channels:
		{
			ret = recorder->numberOfChannels();
			break;
		}
		case dispid_recorder_peakpower:
		{
			ret = recorder->peakPowerForChannel(args[0]->GetIntValue());
			break;
		}
		case dispid_recorder_time:
		{
			ret = recorder->currentTime();
			break;
		}
		case dispid_recorder_input:
		{
			if (args[0]->GetBoolValue()) {
				recorder->input(mp_objs->rwc->inputStream());
			} else {
				recorder->input(mp_objs->input);
			}
			break;
		}

	}
}

void _AudioRecorderDelegate::audioRecorderDidFinishRecording(LPCTSTR file)
{
	CStringW ev;
	ev.Format(L"%s\f%s\f%s", OBJ_NAME, L"stop", CStringW(file));
	mp_intf->postEvent(ev);
}

void _AudioRecorderDelegate::audioRecorderErrorDidOccur(LPCTSTR file)
{
	CStringW ev;
	ev.Format(L"%s\f%s\f%s", OBJ_NAME, L"error", CStringW(file));
	mp_intf->postEvent(ev);
}
